adapters.go

```diff
@@ -212,32 +230,90 @@ func (a *whisperAdapter) RetrieveRaw(filterID string) ([]*whisper.Message, error
 
 	for _, message := range messages {
 		shhMessage := whisper.ToWhisperMessage(message)
-		err := a.decryptMessage(context.Background(), shhMessage)
+		err := a.handleEncryptionMessage(context.Background(), shhMessage)
 		if err != nil {
-			logger.Warn("failed to decrypt a message", zap.Error(err), zap.Binary("messageID", shhMessage.Hash))
+			logger.Warn("failed to handle an encryption message", zap.Error(err), zap.Binary("messageID", shhMessage.Hash))
 		}
 		result = append(result, shhMessage)
 	}
 
 	return result, nil
 }
 
-func (a *whisperAdapter) decodeMessage(message *whisper.Message) (*protocol.StatusMessage, error) {
+func (a *whisperAdapter) decodeMessage(publicKey *ecdsa.PublicKey, message *whisper.Message) (*protocol.StatusMessage, error) {
 
-	publicKey, err := crypto.UnmarshalPubkey(message.Sig)
+	decoded, err := protocol.DecodeMessage(publicKey, message.Payload)
 	if err != nil {
 		return nil, err
 	}
 
-	decoded, err := protocol.DecodeMessage(publicKey, message.Payload)
+	return &decoded, nil
+}
+
+func (a *whisperAdapter) addDatasyncPacket(publicKey *ecdsa.PublicKey, datasyncMessage datasyncproto.Payload) {
+	packet := datasynctransport.Packet{
+		Sender:  datasyncpeer.PublicKeyToPeerID(*publicKey),
+		Payload: datasyncMessage,
+	}
+	a.datasync.AddPacket(packet)
+}
+
+// decodeMessages expects an unencrypted whisper message as it's argument.
+// 1) Check if it's a datasync message
+// 2) If it's a datasync message it will handle each transmitted message separately
+// 3) if it's not a datasync message a single message will be handled
+// Fingerprinting datasync messages is a bit tricky, as they might unmarshal fine
+// because they are wrapped messages, so we need to check whether anything is
+// poulated in the fields.
+func (a *whisperAdapter) decodeMessages(message *whisper.Message) ([]*protocol.StatusMessage, error) {
```

Comment:
Maybe `handleDecodedMessages()`? We can have a flow like `handleEncryptedMessage()` then `handleDecodedMessages()` and `decodeMessage` would stay a low-level message.